src/qobj/functions.jl

```diff
@@ -191,13 +189,17 @@ Quantum Object:   type=Operator   dims=[20, 20]   size=(400, 400)   ishermitian=
 ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠙⠦
 ```
 """
-LinearAlgebra.kron(
-    A::QuantumObject{<:AbstractArray{T1},OpType},
-    B::QuantumObject{<:AbstractArray{T2},OpType},
-) where {T1,T2,OpType<:Union{KetQuantumObject,BraQuantumObject,OperatorQuantumObject}} =
-    QuantumObject(kron(A.data, B.data), A.type, vcat(A.dims, B.dims))
-LinearAlgebra.kron(A::QuantumObject) = A
-function LinearAlgebra.kron(A::Vector{<:QuantumObject})
+function LinearAlgebra.kron(
+    A::AbstractQuantumObject{DT1,OpType},
+    B::AbstractQuantumObject{DT2,OpType},
+) where {DT1,DT2,OpType<:Union{KetQuantumObject,BraQuantumObject,OperatorQuantumObject}}
+    if A isa QuantumObjectEvolution || B isa QuantumObjectEvolution
+        return QuantumObjectEvolution(kron(A.data, B.data), A.type, vcat(A.dims, B.dims))
+    end
+    return QuantumObject(kron(A.data, B.data), A.type, vcat(A.dims, B.dims))
```

Comment:
I think we can extend the basic julia function `promote_type` (maybe put it in the file `utilities.jl` ?) for this.
something like
```julia
promote_type(A::QuantumObjectEvolution, B::QuantumObjectEvolution) = get_typename_wrapper(A)
promote_type(A::QuantumObjectEvolution, B::QuantumObject) = get_typename_wrapper(A)
promote_type(A::QuantumObject, B::QuantumObjectEvolution) = get_typename_wrapper(B)
promote_type(A::QuantumObject, B::QuantumObject) = get_typename_wrapper(A)
```

and therefore, this kind of `if`-condition block can be simplified as
```julia
QType = promote_type(A, B)
QType(kron(A.data, B.data), A.type, vcat(A.dims, B.dims))
```